Fix an issue with invalid popover placement when creating new site

kean opened the pull request:
Fixes https://github.com/wordpress-mobile/WordPress-iOS/issues/23451.

I missed that when testing – it was part of the old code.

<img width="600" alt="Screenshot 2024-07-30 at 10 08 15 AM" src="https://github.com/user-attachments/assets/940ac5e6-cd84-4602-922a-ea403c9714e3">


## Alternatives Considered

I initially implemented it as a nice SwiftUI-based popover with wp.com as a primary button, but you quickly run into issues with SwiftUI and UIKit interop. The popover is presented on top of the site picker as a separate view controller, and there doesn't seem to be any clear way to dismiss it from SwiftUI and guarantee that by the time the site picker tries to present something on top of `presentedViewController`, the popover will actually be dismissed. So you get a "attempting to present on top of the view controller that's already presenting (a popover)" error. Maybe next time.

<img width="600" alt="Screenshot 2024-07-30 at 9 29 22 AM" src="https://github.com/user-attachments/assets/f469891d-2b41-439e-8f0d-d227b5e3fa48">

## Regression Notes
1. Potential unintended areas of impact


2. What I did to test those areas of impact (or what existing automated tests I relied on)


3. What automated tests I added (or what prevented me from doing so)

PR submission checklist:

- [ ] I have completed the Regression Notes.
- [ ] I have considered adding unit tests for my changes.
- [ ] I have considered adding accessibility improvements for my changes.
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

Testing checklist:
- [ ] WordPress.com sites and self-hosted Jetpack sites.
- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] VoiceOver.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] iPhone and iPad. 
- [ ] Multi-tasking: Split view and Slide over. (iPad)

crazytonyli commented on WordPress/Classes/ViewRelated/Blog/Site Picker/SiteSwitcherView.swift:
```diff
@@ -79,9 +81,13 @@ private struct SiteSwitcherView: View {
+                    ZStack {
+                        PopoverSourceView(view: popoverSourceView)
+                        FAB(action: { addSiteAction(popoverSourceView) })
+                    }
```
Out of curiosity, any reason not using SwiftUI [`confirmationDialog`](https://developer.apple.com/documentation/swiftui/view/confirmationdialog(_:ispresented:titlevisibility:presenting:actions:message:)-8y541) to implement this action sheet?

crazytonyli replied:
Sorry, I see the PR description now. You have tried that.

crazytonyli commented:
> So you get a "attempting to present on top of the view controller that's already presenting (a popover)" error.

Do you think we can you use a context menu instead? Is that also a UIKit popover underneath?

kean commented:
> Do you think we can you use a context menu instead? Is that also a UIKit popover underneath?

`Menu` may be a good option. It's not backed by a view controller – at least in the version I tested – so the UIKit presentation after the selection works.

<img width="600px" src="https://github.com/user-attachments/assets/85177ea1-4073-49de-b661-e673cd1bdd59">

I'll have to duplicate or extract some of the code from `AddSiteAlertFactory`, but it's simple enough, so it's not a concern. It's a better option than providing this wacky `PopoverSourceView`. I'll move this PR to draft and update it tmrw.

P.S. There is no way to make one button a "primary" in `Menu`, so it's a bit limiting. I think the ideal target state would be to make the site creation flows self-contained so that they could be presented from SwiftUI.

kean committed:
Fix popover position for create site FAB

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

kean commented on Modules/Sources/DesignSystem/Components/FAB.swift:
```diff
@@ -2,25 +2,34 @@ import SwiftUI
     let image: Image
 
-    @ScaledMetric(relativeTo: .title2) private var size = 50.0
+    @ScaledMetric(relativeTo: .title2) private var size = 54.0
```
Bumped the size a tiny bit because it was smaller than the UIKit version.

kean commented on WordPress/Classes/ViewRelated/Blog/Add Site/AddSiteAlertFactory.swift:
```diff
@@ -49,10 +50,50 @@ class AddSiteAlertFactory: NSObject {
     private func createWPComSiteAction(handler: @escaping () -> Void) -> UIAlertAction {
         return UIAlertAction(
-            title: NSLocalizedString("Create WordPress.com site", comment: "Create WordPress.com site button"),
+            title: Strings.createDotComSite,
```
`AddSiteAlertFactory` is now used only by `BlogListViewController` and will be removed together with it, so I didn't update this implementation.

kean commented on WordPress/Classes/ViewRelated/Blog/Site Picker/SitePickerViewController+SiteActions.swift:
```diff
@@ -25,7 +25,6 @@ extension SitePickerViewController {
         let menuItems = [
             MenuItem.visitSite({ [weak self] in self?.visitSiteTapped() }),
             MenuItem.shareSite { [weak self] in self?.buttonShareSiteTapped() },
-            MenuItem.addSite({ [weak self] in self?.addSiteTapped()} ),
```
This menu is available when long-pressing the site picker dropdown on the dashboard. It's a somewhat hidden feature that was added during a hack week, and we don't officially support it. So, it's safe to remove this option. Almost nobody knows this menu exists. Originally, the plan was to add this menu as a separate button, but it didn't materialize.

kean commented:
@crazytonyli , I updated the implementation to use the context menu, as you suggested. It seems to be simple enough and is unlikely to break. I also dropped a couple of comments about the implementation details.

crazytonyli commented on WordPress/Classes/ViewRelated/Blog/Add Site/AddSiteAlertFactory.swift:
```diff
@@ -49,10 +50,50 @@ class AddSiteAlertFactory: NSObject {
+
+private enum Strings {
+    static let createDotComSite = NSLocalizedString("Create WordPress.com site", comment: "Create WordPress.com site button")
+    static let addSelfHostedSite = NSLocalizedString("Add self-hosted site", comment: "Add self-hosted site button")
```
Use `NSLocalizedString("reverse.domain", value: "...", comment: "...")` here?

kean replied:
Updated. I didn't want to introduce new l10n, but I guess I no reason not to.

crazytonyli commented on Modules/Sources/DesignSystem/Components/FAB.swift:
```diff
@@ -2,25 +2,34 @@ import SwiftUI
 public struct FAB: View {
     let image: Image
-    let action: () -> Void
+    let action: (() -> Void)?
```
It feels pretty strange to have an optional action on a "floating _action_ button" type... Also, I didn't see a place that actually use a nil action (there is one `FAB(action:{})` though). What do you think reverting it back to non-optional?

kean replied:
See https://github.com/wordpress-mobile/WordPress-iOS/pull/23454/files#r1700425910. I'm not entirely happy with this API design, but I don't think it's worth changing. 

We've been discussing removing FABs soon anyway 👀

crazytonyli commented on WordPress/Classes/ViewRelated/Blog/Site Picker/SiteSwitcherView.swift:
```diff
@@ -66,12 +66,32 @@ private struct SiteSwitcherToolbarView: View {
+                    Button(action.title, action: action.handler)
+                }
+            } label: {
+                FAB(action: {})
```
I presume you want the content, but not the action button in `FAB`? It'd be great if we can make the content view available for use. I noticed there seems to be some complications around dynamic type. Maybe that's a blocker in publishing the `FABContenxtView`?

kean replied:
Oh, I added a variant that doesn't require an action, but forgot to use it – updated.

> I noticed there seems to be some complications around dynamic type. Maybe that's a blocker in publishing the FABContentView?

It's a similar situation as with `@Environment(isSearching`. `@ScaledMetric` has to be _inside_ the view that sets `dynamicTypeSize`. So there has a to be a wrapper for `FABContentView`. I think using `FAB` for this purpose is OK.
